File: 2023/01/main.go
package main

import (
	"fmt"
	"io"
	"os"
	"regexp"
	"strings"
)

func main() {
	f, err := os.Open("input.txt")
	if err != nil {
		panic(err)
	}

	contents, err := io.ReadAll(f)
	if err != nil {
		panic(err)
	}

	regexes := []*regexp.Regexp{
		regexp.MustCompile(`(\d|one|three|four|five|six|seven)`),
		regexp.MustCompile(`(eight|nine)`),
		regexp.MustCompile(`(two)`),
	}

	m := map[string]int{
		"zero":  0,
		"0":     0,
		"1":     1,
		"one":   1,
		"2":     2,
		"two":   2,
		"3":     3,
		"three": 3,
		"4":     4,
		"four":  4,
		"5":     5,
		"five":  5,
		"6":     6,
		"six":   6,
		"7":     7,
		"seven": 7,
		"8":     8,
		"eight": 8,
		"9":     9,
		"nine":  9,
	}

	var total int
	lines := strings.Split(string(contents), "\n")
	for _, line := range lines {
		if line == "" {
			continue
		}

		firstStart := len(line)
		firstEnd := len(line)
		lastStart := -1
		lastEnd := -1
		for _, re := range regexes {
			indices := re.FindAllStringSubmatchIndex(line, -1)
			if len(indices) == 0 {
				continue
			}

			f := indices[0]
			if f[0] < firstStart {
				firstStart = f[0]
				firstEnd = f[1]
			}

			l := indices[len(indices)-1]
			if l[0] > lastStart {
				lastStart = l[0]
				lastEnd = l[1]
			}

		}

		first := m[line[firstStart:firstEnd]]
		last := m[line[lastStart:lastEnd]]

		sub := (first * 10) + last
		fmt.Printf("%v: %v\n", line, sub)

		total += sub
	}

	fmt.Printf("total: %v\n", total)
}

File: 2023/01/main.zig
const std = @import("std");

pub fn main() !void {
    // Open input file.
    var file = try std.fs.cwd().openFile("input.txt", .{ .mode = .read_only });
    defer file.close();

    // Get file info.
    var file_stat = try file.stat();

    // Get file size.
    const file_size = file_stat.size;

    // Initialize allocator.
    const allocator = std.heap.page_allocator;

    // Read file contents.
    const buffer = try file.readToEndAlloc(allocator, file_size);
    defer allocator.free(buffer);

    var map = std.StringHashMap(u8).init(allocator);
    try map.put("1", 1);
    try map.put("one", 1);
    try map.put("2", 2);
    try map.put("two", 2);
    try map.put("3", 3);
    try map.put("three", 3);
    try map.put("4", 4);
    try map.put("four", 4);
    try map.put("5", 5);
    try map.put("five", 5);
    try map.put("6", 6);
    try map.put("six", 6);
    try map.put("7", 7);
    try map.put("seven", 7);
    try map.put("8", 8);
    try map.put("eight", 8);
    try map.put("9", 9);
    try map.put("nine", 9);
    defer map.deinit();

    const candidates = [_][]const u8{
        "1",
        "one",
        "2",
        "two",
        "3",
        "three",
        "4",
        "four",
        "5",
        "five",
        "6",
        "six",
        "7",
        "seven",
        "8",
        "eight",
        "9",
        "nine",
    };

    var line = std.ArrayList(u8).init(allocator);
    defer line.deinit();
    var lines = std.ArrayList(std.ArrayList(u8)).init(allocator);
    defer lines.deinit();

    for (buffer) |character| {
        if (character == '\n') {
            try lines.append(line);
            line = std.ArrayList(u8).init(allocator);
            continue;
        }

        try line.append(character);
    }

    var total: u32 = 0;
    for (lines.items) |l| {
        std.debug.print("line: {s}\n", .{l.items});
        var list = std.ArrayList([]const u8).init(allocator);
        defer list.deinit();
        for (0..l.items.len) |i| {
            for (candidates) |candidate| {
                if (l.items.len - i >= candidate.len and std.mem.eql(u8, candidate, l.items[i .. i + candidate.len])) {
                    std.debug.print("c: {s}: {d}\n", .{ candidate, l.items.len });
                    try list.append(candidate);
                }
            }
        }

        var first = map.get(list.items[0]).?;
        var last = map.get(list.items[list.items.len - 1]).?;
        total += first * 10 + last;
    }

    std.debug.print("Total {d}\n", .{total});
}
